install-gateway.ps1

```diff
@@ -1,79 +1,144 @@
 
-# Wait 120 Seconds for the installation to finish
-    Start-Sleep -s 180
+# Wait for the installation to finish
+  Write-Host 'Waiting for jreinstall.exe to finish...'
+  while ((Get-Process jreinstall -ErrorAction SilentlyContinue))
+  {
+    Start-Sleep -Seconds 10
+  }
 
 # Remove the installer
-    rm -Force $workd\jre*
+  rm -Force $workd\jre*
+
+  Print-ExecInfo $process
 }
 
-function Install-VcRedist()
+function Install-VcRedist([string]$vcrsource = $script:vcrsource, [string]$vcrsource_filter = $script:vcrsource_filter)
 {
-  $workd = "c:\temp"
-    If (!(Test-Path -Path $workd -PathType Container))
-    {
-      New-Item -Path $workd -ItemType directory
-    }
-  $destination = "$workd\vcredist_x64.exe"
-    $client = New-Object System.Net.WebClient
-    $client.DownloadFile("https://download.microsoft.com/download/3/2/2/3224B87F-CFA0-4E70-BDA3-3DE650EFEBA5/vcredist_x64.exe", $destination)
-    Start-Process -FilePath "$workd\vcredist_x64.exe" -ArgumentList "/Q" -Wait
+  $filename = Download-File -uri $vcrsource -msdl_filter $vcrsource_filter
+
+  Write-Verbose "Installing VcRedist from $filename"
+  $process = Start-Process -FilePath "$filename" -ArgumentList "/Q /log $workd\vcredist.log" -Wait -PassThru
+  
+  Start-Sleep 2
+  Print-ExecInfo $process
 }
 
-function Install-Gateway([string] $gwPath)
+function Install-Gateway([string]$filename = $script:path, [string]$workd = $script:workd)
 {
 # uninstall any existing gateway
   UnInstall-Gateway
 
-    Write-Host "Start Gateway installation"
+  Write-Host 'Start Gateway installation'
+  
+  Write-Verbose "Installing Gateway from $filename"
+  
+  $process = Start-Process -FilePath "msiexec.exe" -ArgumentList "/i $filename /quiet /passive /L*v $workd\ir.log" -Wait -PassThru
+  Start-Sleep -Seconds 30
 
-    Start-Process "msiexec.exe" "/i $path /quiet /passive" -Wait
-    Start-Sleep -Seconds 30
+  Write-Host "Gateway installer finished with exitCode: $($process.exitcode))"
 
-    Write-Host "Succeed to install gateway"
+  Print-ExecInfo $process
 }
 
 function Register-Gateway([string] $key)
 {
   Write-Host "Start to register gateway with key: $key"
-    & "C:\Program Files\Microsoft Integration Runtime\5.0\Shared\dmgcmd.exe" @("-EnableRemoteAccess", "8060")
-    & "C:\Program Files\Microsoft Integration Runtime\5.0\Shared\dmgcmd.exe" @("-RegisterNewNode", "$key", "$env:COMPUTERNAME")
-    Write-Host "Succeed to register gateway"
+  $executable = "C:\Program Files\Microsoft Integration Runtime\5.0\Shared\dmgcmd.exe"
+  
+  if (test-path $executable)
+  {
+    $p1 = Start-Process -FilePath $executable -ArgumentList @("-EnableRemoteAccess", "8060") -Wait -PassThru
+    $p2 = Start-Process -FilePath $executable -ArgumentList @("-RegisterNewNode", "$key", "$env:COMPUTERNAME") -Wait -PassThru
+  }
+
+  switch($p1.exitCode + $p2.exitCode)
+  {
+    0
+      {
+        Write-Host "Succeed to register the gateway"
```

Comment:
```suggestion
        Write-Host "Registration with the gateway was successul"
```

Reply:
I thought on removing all the write-host blocks altogether, as now there is a special function to handle operation results with fair amount of information. I should have posted a sample output for a reference.